rfcs/bring-ts.md

```diff
@@ -0,0 +1,413 @@
+
+OK, let's use this library in our Wing code.
+
+```js
+bring colors;
+
+print("${colors.Global.green("hello")}, ${colors.Colors.blue("world")}!")
+```
+
+As you can see, since Wing doesn't support global functions, the `Global` wrapper class was
+generated and includes a static method for each public API exported from `colors`.
+
+Similarly to turning global functions to static methods, `ts2wing` applies other heuristics to
+create an [adaptation layer](#type-system-adapters) between the type systems.
+
+### Escape hatch
+
+As you might have expected, there could be cases where `ts2wing` won't be able to create an adapter
+for a certain TypeScript library. It will take us time to establish all the patterns and in parallel
+evolve the Wing type system.
+
+If you run `ts2wing` with the switch `--unpacked` like so:
+
+```sh
+$ ts2wing edit colors
+OK, "colors" is ready to be edited under "vendor/colors.wlib.src".
+After you are done, make sure to run: ts2wing colors
+```
+
+Now, go ahead and make any modifications you like to `vendor/colors.wlib.src` and after
+you are finished, run:
+
+```sh
+$ ts2wing colors
+Found vendor/colors.wlib.src
+Processing type definitions (index.d.ts)...
+Generating Wing wrapper classes...
+Compiling with JSII...
+Generating API.md...
+Packaging Wing Library to vendor/colors.wlib
+Cleaning vendor/colors.wlib.src
+```
+
+This is it, now the updated `colors.wlib` will include your changes.
+
+### Type system adapters
+
+It is not possible to express every type in TypeScript within the Wing type system. This is because:
+
+1. Wing is young and it will take the type system time to evolve; 
+2. The design goals of Wing and TypeScript are different. TypeScript is designed to be able to
+   represent every possible type that can be expressed in JavaScript, while Wing is designed to be a
+   great language for cloud application programming.
+
+#### Union Types
+
+A good example is **type unions**. Since in JavaScript is dynamic, it is possible for a value to
+have multiple possible types. In TypeScript this is expressed through unions.
+
+Let's take an example:
+
+```ts
+declare class Foo {
+  constructor(value: string | number);
+  bar():  string[] | string | false;
+}
+```
+
+The constructor of `Foo` accepts a single argument that can be either a `string` or a `number`.
+
+This is all nice, but at the moment we believe that supporting type unions is not really required
+for building great cloud applications. In fact, we think that type unions are a potential source of
+bugs and complexity, and we want Wing to be easy to learn and use and scale across large teams in
+big companies. To that end, we don't support type unions (and we reserve the right to change our
+minds).
+
+To bridge this gap, the `ts2wing` tool will generate a "union-like class" in for each type union in
+the imported library. The class will have a set of static factory methods, one for each type in the
+union:
+
+```ts
+declare class FooInitValue {
```

Comment:
I am trying to imagine the actual `bridge.ts` artifact in this example, hope I understand this correctly.
Also, this is the first time in my life that I am writing typescript code, and I have to say, It is Fucking amazing!!! 

So, bare with me cause I am a noob in typescript...

Imagine we have the following class Foo implementation:

```ts
// foo.ts
export class Foo {
  private value: string | number;
  constructor(value: string | number) {
    console.log(`The parameter is of type ${typeof value}`);
    this.value = value;
  }

  bar(): string[] | string | false {
    if (Math.random() > 0.5) return ["hello", "world"];
    if (Math.random() > 0.5) return "hello world";
    return false;
  }

  toString() {
    return `This is a Foo Class with value ${this.value}`;
  }

  getAnotherFoo(f: Foo) {
    console.log(`got another foo ${f}`);
  }
}
```

I assume we will need to result in a `foo_bridge.ts` file
```ts
// foo_bridge.ts
import { ExampleClass, globalExample, Foo as _inner_Foo } from "./original";

export class FooInitValue {
  static fromStr(value: string): FooInitValue {
    return new FooInitValueImp(value, undefined);
  }
  static fromNum(value: number): FooInitValue {
    return new FooInitValueImp(undefined, value);
  }
}
class FooInitValueImp {
  _str?: string;
  _number?: number;
  
  constructor(s?: string , n?: number){
    if (s) this._str = s; 
    else if (n) this._number = n;
    else throw new Error("Unreachable");
    
  }
  
}
export declare class FooBarResult{
  asBool(): boolean ;
  isBool(): boolean ;
  asMutArrayStr(): string[];
  isMutArrayStr(): boolean ;
  asString(): string ;
  isString(): boolean ;
}
class FooBarResultImpl implements FooBarResult {
  _boolean? : boolean;
  _mutArrayStr?: string[];
  _string?: string;
  asBool(): boolean {
    return this._boolean!;
  }
  isBool(): boolean {
    return typeof this._boolean !== "undefined";
  }
  isMutArrayStr(): boolean {
    return typeof this._mutArrayStr !== "undefined";
  }
  asMutArrayStr(): string[] {
    return this._mutArrayStr!;
  }
  isString(): boolean {
    return typeof this._string !== "undefined";
  }
  asString(): string {
    return this._string!;
  }
}

export class Foo {
  _boxed: _inner_Foo;
  constructor(v: FooInitValue){
    if (v instanceof FooInitValueImp){
      if (v._number) this._boxed = new _inner_Foo(v._number);
      else if (v._str)  this._boxed = new _inner_Foo(v._str);
      else throw new Error("Unreachable");
    }
    else {
      throw new Error("Unreachable");
    }
  }

  toString(){
    return this._boxed.toString();
  }

  bar():  FooBarResult {
    let innerRes = this._boxed.bar();
    let res = new FooBarResultImpl();
    if (typeof innerRes ===  "string" ) res._string = innerRes;
    else if (innerRes instanceof  Array) res._mutArrayStr = innerRes;
    else {
      res._boolean = innerRes;
    }
    return res;
  }
  getAnotherFoo(f: Foo ){
    return this._boxed.getAnotherFoo(f._boxed);
  }
}
```

And then the consumer of the bridge class will look like this:
```ts
import { Foo, FooBarResult, FooInitValue} from "./bridge";

let fooFromNumber = new Foo(FooInitValue.fromNum(10));
let fooFromString = new Foo(FooInitValue.fromStr("ten"));
fooFromNumber.getAnotherFoo(fooFromString);
let bar = new Foo(FooInitValue.fromNum(20)).bar();

if (bar.isBool()) {
  console.log(`bar ${bar} is bool = ${bar.asBool()}`);
}
if (bar.isMutArrayStr()) {
  console.log(`bar ${bar} is MutArrayStr = ${bar.asMutArrayStr()}`);
}
if (bar.isString()) {
  console.log(`bar ${bar} is String = ${bar.asString()}`);
}
```

First, is this correct? or close to being correct? 
Second, please notice that any exported entity that is linked to the exported original Foo also needs to be implemented in the bridge using the new `BoxedFoo`